File: bindings/python/src/datalevin/entity.py
"""Lazy Datalevin entity wrapper."""

from __future__ import annotations

from collections.abc import Mapping

import jpype
from jpype.types import JByte

from ._convert import to_python
from ._interop import _BINDINGS
from ._java import call_java, classes, is_java_object

_MISSING = object()


def _entity_value(value):
    if value is None or isinstance(value, (bool, int, float, str, bytes)):
        return value

    if not is_java_object(value):
        if isinstance(value, Mapping):
            return {_entity_value(key): _entity_value(item) for key, item in value.items()}
        if isinstance(value, (set, frozenset)):
            return {_entity_value(item) for item in value}
        if isinstance(value, (list, tuple)):
            return [_entity_value(item) for item in value]
        return value

    if _BINDINGS.entity_is(value):
        return Entity(value)

    cls = classes()
    byte_array_type = jpype.JArray(JByte)

    if isinstance(value, byte_array_type):
        return bytes(value)

    if isinstance(value, cls.map_type) or hasattr(value, "entrySet"):
        result = {}
        for entry in value.entrySet():
            result[to_python(entry.getKey())] = _entity_value(entry.getValue())
        return result

    if isinstance(value, cls.set_type):
        return {_entity_value(item) for item in value}

    if isinstance(value, (cls.list_type, cls.collection_type)) or hasattr(value, "iterator"):
        return [_entity_value(item) for item in value]

    return to_python(value)


class Entity(Mapping):
    """Lazy entity returned by `Connection.entity()`."""

    def __init__(self, handle) -> None:
        self._handle = handle

    def raw_handle(self):
        return self._handle

    @property
    def id(self):
        """Return the Datalevin entity id."""

        return to_python(_BINDINGS.entity_id(self._handle))

    def get(self, attr, default=None):
        if not self.__contains__(attr):
            return default
        return _entity_value(_BINDINGS.entity_get(self._handle, attr))

    def touch(self):
        """Materialize all current entity attributes into a Python dict."""

        return to_python(_BINDINGS.entity_touch(self._handle))

    def __getitem__(self, attr):
        value = self.get(attr, _MISSING)
        if value is _MISSING:
            raise KeyError(attr)
        return value

    def __contains__(self, attr):
        return _BINDINGS.entity_contains(self._handle, attr)

    def __iter__(self):
        return iter(self.touch())

    def __len__(self):
        return len(self.touch())

    def __eq__(self, other):
        if not isinstance(other, Entity):
            return NotImplemented
        return bool(call_java(self._handle.equals, other.raw_handle()))

    def __hash__(self):
        return int(call_java(self._handle.hashCode))

    def __repr__(self) -> str:
        return "<Entity lazy>"

File: bindings/javascript/src/entity.js
import { isJavaObject, toJs } from "./convert.js";
import { _BINDINGS } from "./interop.js";
import { classes } from "./jvm.js";

function materializeJavaCollection(collection) {
  if (typeof collection?.toArraySync === "function") {
    try {
      const items = collection.toArraySync();
      return Array.isArray(items) ? items : Array.from(items);
    } catch {
      // Fall back to iterator-based traversal when array materialization is unsupported.
    }
  }

  const items = [];
  const iterator = collection.iteratorSync();
  while (iterator.hasNextSync()) {
    items.push(iterator.nextSync());
  }
  return items;
}

function instanceOf(value, type) {
  if (value === null || typeof value !== "object" || typeof value.instanceOf !== "function") {
    return false;
  }

  try {
    return value.instanceOf(type);
  } catch {
    return false;
  }
}

function isPlainObject(value) {
  if (value === null || typeof value !== "object") {
    return false;
  }
  const prototype = Object.getPrototypeOf(value);
  return prototype === Object.prototype || prototype === null;
}

async function entityValue(value) {
  if (
    value === null
    || typeof value === "boolean"
    || typeof value === "number"
    || typeof value === "string"
    || typeof value === "bigint"
  ) {
    return value;
  }

  if (Array.isArray(value)) {
    return Promise.all(value.map((item) => entityValue(item)));
  }

  if (!isJavaObject(value)) {
    if (value instanceof Set) {
      const items = [];
      for (const item of value) {
        items.push(await entityValue(item));
      }
      return new Set(items);
    }
    if (value instanceof Map) {
      const entries = [];
      for (const [key, item] of value) {
        entries.push([await entityValue(key), await entityValue(item)]);
      }
      return new Map(entries);
    }
    if (isPlainObject(value)) {
      const result = {};
      for (const [key, item] of Object.entries(value)) {
        result[key] = await entityValue(item);
      }
      return result;
    }
    return value;
  }

  if (await _BINDINGS.entityIs(value)) {
    return new Entity(value);
  }

  const cls = await classes();

  if (instanceOf(value, cls.mapType) || typeof value.entrySetSync === "function") {
    const entries = [];
    for (const entry of materializeJavaCollection(value.entrySetSync())) {
      entries.push([await toJs(entry.getKeySync()), await entityValue(entry.getValueSync())]);
    }

    if (entries.every(([key]) => typeof key === "string")) {
      return Object.fromEntries(entries);
    }
    return new Map(entries);
  }

  if (instanceOf(value, cls.setType)) {
    const items = [];
    for (const item of materializeJavaCollection(value)) {
      items.push(await entityValue(item));
    }
    return new Set(items);
  }

  if (
    instanceOf(value, cls.listType)
    || instanceOf(value, cls.collectionType)
    || typeof value.iteratorSync === "function"
  ) {
    const items = [];
    for (const item of materializeJavaCollection(value)) {
      items.push(await entityValue(item));
    }
    return items;
  }

  return toJs(value);
}

export class Entity {
  constructor(handle) {
    this.handle = handle;
  }

  rawHandle() {
    return this.handle;
  }

  async id() {
    return toJs(await _BINDINGS.entityId(this.rawHandle()));
  }

  async has(attr) {
    return _BINDINGS.entityContains(this.rawHandle(), attr);
  }

  async get(attr, defaultValue = undefined) {
    if (!(await this.has(attr))) {
      return defaultValue;
    }
    return entityValue(await _BINDINGS.entityGet(this.rawHandle(), attr));
  }

  async touch() {
    return toJs(await _BINDINGS.entityTouch(this.rawHandle()));
  }

  async entries() {
    const value = await this.touch();
    if (value instanceof Map) {
      return Array.from(value.entries());
    }
    return Object.entries(value ?? {});
  }

  async keys() {
    return (await this.entries()).map(([key]) => key);
  }

  async values() {
    return (await this.entries()).map(([, value]) => value);
  }

  toString() {
    return "<Entity lazy>";
  }
}
